### PR TITLE
Cache query results in the simple container

### PR DESCRIPTION
**Pull request by rullzer**
More generic caching approach of suggested cache by @icewind1991 in https://github.com/owncloud/core/pull/21419

@icewind1991 can you verify with the same setup that this gives the same speedup?

@BernhardPosselt @MorrisJobke Please have a look. This now caches (cacheable services) for all simplecontainers.

### CODE REVIEW

**Comment by Xenopathic** [comment]
Isn't this just what `offsetGet()` does anyway?

**Comment by rullzer** [comment]
@Xenopathic yeah but slighly less efficient apparently if you look at the numbers in https://github.com/owncloud/core/pull/21419

**Comment by Xenopathic** [comment]
Then we should fix Pimple :smile:

The code at the moment basically looks like this:

```
if (isCached($object)) {
  return $cache[$object];
}
if (isCached2($object)) {
  return $cache2[$object];
}
return constructObject($object);
```

which is a little silly...

File: lib/private/appframework/utility/simplecontainer.php [hunk 1]
```diff
@@ -130,12 +147,15 @@ public function registerService($name, Closure $closure, $shared = true) {
...
 			unset($this[$name]);
+			unset($this->cache[$name]);
+			unset($this->cacheable[$name]);
 		}
 		if ($shared) {
 			$this[$name] = $closure;
 		} else {
 			$this[$name] = parent::factory($closure);
 		}
+		$this->cacheable[$name] = $shared;
```
Comment (nickvergessen) [inline lib/private/appframework/utility/simplecontainer.php]: This should be inside the `if ($shared)` as stated in the other pr

Comment (nickvergessen) [inline lib/private/appframework/utility/simplecontainer.php]: Ah you store true and false, so should be fine

Comment (rullzer) [inline lib/private/appframework/utility/simplecontainer.php]: Yeah I figured maybe at some point we really want to know if it is explicitly marked as not shared.

**Comment by rullzer** [comment]
> Then we should fix Pimple :smile:

@Xenopathic I do agree. But the caching in Pimple is slightly different. They have a lot more safety checks (which probably what makes it slower). Since Pimple does that for us we do not need to in our cache.

But fixing Pimple is cleaner indeed.

**Comment by rullzer** [comment]
So after thinking about it some more today I agree with @Xenopathic we should provide patches to Pimple. Let me close this.

**Comment by DeepDiver1975** [comment]
> So after thinking about it some more today I agree with @Xenopathic we should provide patches to Pimple. Let me close this.

@rullzer did you open a PR on this one with Pimple? THX

**Comment by rullzer** [comment]
Not yet. Have to look into pimple a bit more.. it is on the list :)

**Pull request closed without merging**